drivers: power: sam: Introduce Atmel SAM SUPC, wakeup sources and poweroff implementation

bjarki-andreasen opened the pull request:
This PR adds two features required for power off of the atmel sam series chips
* Driver and DTS entry for SUPC. This component sets up wakeup sources and powers down the core voltage regulator, which is required to go to sleep and wake up again.
* sys_poweroff implementation required to synchronously enable the wakeup sources using SUPC, and entering deep sleep (backup mode)

gmarull commented on soc/arm/atmel_sam/common/soc_poweroff.c:
```diff
@@ -0,0 +1,19 @@
+ */
+
+#include <zephyr/sys/poweroff.h>
+#include <soc.h>
+
+#ifndef SUPC_CR_KEY_PASSWD
+#define SUPC_CR_KEY_PASSWD (0xA5 << 24)
+#endif
```
usase of ifndef here seems to hide some header issues, can you please explain?

bjarki-trackunit replied:
The headers for the atmel sam chips in `modules/hal/atmel/asf` are irregular, defining the SUPC_CR_KEY_PASSWD only for some chips, this guard ensures the value is not redefined, and is included in this form for the remaining chips.

Examples of defines from different headers for the same hardware:
```
#define SUPC_CR_KEY(value) ((SUPC_CR_KEY_Msk & ((value) << SUPC_CR_KEY_Pos)))
```
```
#define   SUPC_CR_KEY_PASSWD (0xA5u << 24)
```

gmarull replied:
does this value depend on SoC?

bjarki-trackunit replied:
No, its constant :) Just wanted to reuse the name for clarity, I could just not use it

gmarull replied:
If HAL is inconsistent I'd just define it with a similar name locally and forall, so no need for ifndef hackery.

bjarki-trackunit replied:
Hacked off ifndef hackery

nandojve commented on soc/arm/atmel_sam/common/soc_poweroff.c:
```diff
@@ -0,0 +1,17 @@
+void z_sys_poweroff(void)
+{
+	/* Switch off voltage regulator to enter backup mode */
+	SUPC->SUPC_CR = SUPC_CR_KEY_PASSWORD | SUPC_CR_VROFF_STOP_VREG;
```
As stated at [hal_atmel/repository-specific-guidelines](https://github.com/zephyrproject-rtos/hal_atmel#repository-specific-guidelines) fixups should be send to hal.

Please, follow the guidelines at hal_atmel to send missing fixups and start to use
```diff
-	SUPC->SUPC_CR = SUPC_CR_KEY_PASSWORD | SUPC_CR_VROFF_STOP_VREG;
+	SUPC->SUPC_CR = SUPC_CR_KEY(SUPC_CR_KEY_PASSWD);
```

This is a partial implementation and not solve the other case that is when user design using the external power supply for VDDCODE. My initial reading is that a SUPC dts/driver should be added. I think at end you will need select the WKP sources too and maybe manage other topics like RTC.

I'm open to suggestion about this because it is not clear to me what kind of dts binding this probably will be. I don't expect a full driver and I believe in this case is start with the bare minimal and than incrementally add features.

bjarki-trackunit replied:
We could add it like
```
		supc: supc@400e1810 {
			compatible = "atmel,sam-supc";
			reg = <0x400e1810 0xd4>;
		};
```

and create some basic private APIs for entering backup mode and enabling gpio and component (RTC) wakeup events.

Then we can address it from the RTC driver if it is set as a `wakeup-source` (and to get the address of the SYSC_WPMR register)

```
		rtc: rtc@400e1810 {
			compatible = "atmel,sam-rtc";
			reg = <0x400e1810 0xd4>;
			supc = <&supc>;
		};
```

bjarki-trackunit replied:
What to do with GPIO wakeup events I'm not sure off yet. Maybe the gpio-controller could address the supc as well, and enable wakeup if we set some custom flag for the GPIO, like `ATMEL_SAM_WAKEUP` or something. but thats for later

nandojve replied:
I think we can go the way you are proposing but let's keep GPIO WKP pins outside for now.
Do you agreed that there is space inside supc for sys_poweroff ?

bjarki-trackunit replied:
not exactly inside, sys_poweroff would call SUPC regulator VROFF (something like `supc_vreg_off()`), or the CMSIS DEEPSLEEP directly for the chips that require that :)

bjarki-trackunit commented:
@nandojve I have created a [PR](https://github.com/zephyrproject-rtos/hal_atmel/pull/33) to add missing defines for registers required by this PR and the RTC PR (to avoid some ugly hackery), please review it :)

Commit:
soc: sam: Add SUPC driver and dts model

This commit adds a driver and dts model for the ATMEL SAM
SUPC component.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Commit:
dts: soc: atmel: sam: Add SUPC component to soc dtsi

This commit adds the new SUPC devicetree instance to the
soc dtsi files.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

Commit:
soc: sam: Add poweroff implementation

This commit adds an implementation of poweroff, which first
uses SUPC to enable all defined wakeup sources (except for
sam4l), followed by entering backup mode.

Signed-off-by: Bjarki Arge Andreasen <bjarkix123@gmail.com>

bjarki-trackunit commented:
@nandojve Please revisit this PR when you find the time :) It is now a quite complete solution, supporting enabling wakeup sources through devicetree through the SUPC component :)

nandojve commented:
> @nandojve Please revisit this PR when you find the time :) It is now a quite complete solution, supporting enabling wakeup sources through devicetree through the SUPC component :) A quick example
> 
> ```
> &supc {
> 	enable-wakeup-source-rtt;
> 	enable-wakeup-source-rtc;
> 	enabled-wakeup-inputs-mask = <0x3>;
> 	enabled-wakeup-inputs-type-mask = <0x1>;
> };
> ```
> 
> Later down the line, we can create a soc specific headers with defines for the masks, so we can do something like
> 
> ```
> 	enabled-wakeup-inputs-mask = < SAM_SUPC_WKUP_PA11 |
>                                        SAM_SUPC_WKUP_PB2 >;
> ```

Hi @bjarki-trackunit ,

I would like to present a little bit different view about this.
I have the impression that we should have a devicetree consumer node like

```
&rtc {
    something = <&supc SUPC_RTC_WAKEUP_SOURCE>;
    wake-up; /* tells that RTC should configure itself as a wake-up source using supc API */
or
    something = <&supc SUPC_RTC_WAKEUP_SOURCE ACTIVE>;
};
```

In this case, supc driver exposes an API that only enable/disable the wake-up sources like RTC as a parameter which can be used by consumers. Try think that RTC wake-up feature could be handled "automatically" by the PM subsystem. If SUPC should be only configured at system init it doesn't change the SUPC API but how driver will register/unregister itself as a wake-up source. This way SUPC devicetree node doesn't have any special property.

The wake-up GPIO pins will use a similar approach. The difference is that it is handled by the pinctrl driver using the SUPC API to enable/disable the source pin.

> SAM_SUPC_WKUP_PA11

This was already defined at hal but I realized that maybe it will require a small change. This is way I'm asking you to postpone everything related to the wake-up pins.

What I'm telling you makes sense? Do you see issues / drawbacks?

gmarull reviewed (changes requested):
-1 to custom wake-up mechanisms, things should be standardized (see e.g. `wakeup-source` dt prop)

nandojve commented:
> -1 to custom wake-up mechanisms, things should be standardized (see e.g. `wakeup-source` dt prop)

Hi @gmarull ,

I was wondering if you can clarify better because I don't know if you are referring to @bjarki-trackunit proposal or my comments. It will be nice if you can add here, to be clear, what is the standard too so we can follow you better about `wakeup-source` dt prop.

gmarull commented:
> > -1 to custom wake-up mechanisms, things should be standardized (see e.g. `wakeup-source` dt prop)
> 
> Hi @gmarull ,
> 
> I was wondering if you can clarify better because I don't know if you are referring to @bjarki-trackunit proposal or my comments. It will be nice if you can add here, to be clear, what is the standard too so we can follow you better about `wakeup-source` dt prop.

There's no standard, but I want to avoid more vendor-specific stuff proliferating in the tree. We now have 1 sample per vendor to accomplish the same thing: wake-up from an X source. This needs to be standardized. The DT `wakeup-source` flag may be a starting point.

bjarki-trackunit commented:
@nandojve @gmarull Using the RTC device as an example, we could then have
```
rtc: rtc@01010101 {
        ...
        wakeup-source-channel = <&supc SUPC_CHANNEL_RTC_ALARM>;
};
```
in the .dtsi and
```
&rtc: {
        ...
        wakeup-source;
        status = "okay";
};
```
in the overlay :)

gmarull commented:
> @nandojve @gmarull Using the RTC device as an example, we could then have
> 
> ```
> rtc: rtc@01010101 {
>         ...
>         wakeup-source-channel = <&supc SUPC_CHANNEL_RTC_ALARM>;
> };
> ```
> 
> in the .dtsi and
> 
> ```
> &rtc: {
>         ...
>         wakeup-source;
>         status = "okay";
> };
> ```
> 
> in the overlay :)

lgtm

nandojve commented:
Hi @bjarki-trackunit ,

> @nandojve @gmarull Using the RTC device as an example, we could then have
> 
> ```
> rtc: rtc@01010101 {
>         ...
>         wakeup-source-channel = <&supc SUPC_CHANNEL_RTC_ALARM>;
> };
> ```

Let's use `supc` for now since it have multiple functions and it will be used in more places.

 ```
 rtc: rtc@01010101 {
         ...
         supc = <&supc SUPC_CHANNEL_RTC_ALARM>;
or
         atmel,supc = <&supc SUPC_CHANNEL_RTC_ALARM>;
 };
 ```

The below seems to be the correct way to enable the wake-up source.

> in the .dtsi and
> 
> ```
> &rtc: {
>         ...
>         wakeup-source;
>         status = "okay";
> };
> ```

bjarki-trackunit commented:
> Let's use `supc` for now since it have multiple functions and it will be used in more places.
> 
> ```
> rtc: rtc@01010101 {
>         ...
>         supc = <&supc SUPC_CHANNEL_RTC_ALARM>;
> or
>         atmel,supc = <&supc SUPC_CHANNEL_RTC_ALARM>;
> };
> ```

I believe the `wakeup-source-channel` is a configuration the "child" device driver will use to enable itself as a wakeup source, something that would expand into something like
```
uint32_t wakeup_source_channel = SAM_DT_INST_SUPC_WAKEUP_CHANNEL(inst);
```
which will then be called from the child device driver like
```
soc_supc_enable_wakeup_source(uint32_t wakeup_source_channel);
```

Am I misunderstanding what the property is supposed to be? seems like `wakeup-source-channel` is a more apt property name than `supc`?

nandojve commented:
Hi @bjarki-trackunit ,

You are right in your approach defining the #wakeup-source-id-cells at supc. Let's use it.

nandojve commented on soc/arm/atmel_sam/common/CMakeLists.txt:
```diff
@@ -3,6 +3,9 @@
 
 zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_SAM4L soc_sam4l_pm.c)
 zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_SAM4L soc_sam4l_gpio.c)
+zephyr_library_sources_ifndef(CONFIG_SOC_SERIES_SAM4L soc_poweroff.c)
```
you mean

```diff
-zephyr_library_sources_ifndef(CONFIG_SOC_SERIES_SAM4L soc_poweroff.c)
+zephyr_library_sources_ifdef(CONFIG_SOC_SERIES_SAM4L soc_sam4l_poweroff.c)
```

bjarki-trackunit replied:
fixed

nandojve commented on soc/arm/atmel_sam/common/soc_supc.h:
```diff
@@ -0,0 +1,27 @@
+#define _ATMEL_SAM_SOC_SUPC_H_
+
+#include <zephyr/types.h>
+#include <soc.h>
```
This include is not necessary.

bjarki-trackunit replied:
removed

bjarki-trackunit commented:
The failing test will succeed once the hal is updated :)

nandojve commented:
Hi @bjarki-trackunit ,

I was wondering if you can update this PR to us check if is already passing CI.

bjarki-trackunit commented on dts/bindings/power/atmel,sam-supc.yaml:
```diff
@@ -0,0 +1,47 @@
+  Atmel SAM SUPC (Supply-Controller) controller
+
+  The supply controller manages wakeup sources and core voltages and
+  regulators.
```
Replaced with your suggestion

bjarki-trackunit commented on dts/bindings/power/atmel,sam-supc.yaml:
```diff
@@ -0,0 +1,47 @@
+
+  To specify the wakeup source channel used by a peripheral, the custom
+  property wakeup-source-channel needs to be used, e.g. for the RTC
+  device:
```
Replaced with your suggestion

bjarki-trackunit commented on dts/bindings/power/atmel,sam-supc.yaml:
```diff
@@ -0,0 +1,47 @@
+
+  The device drivers will then use the provided channel to enable its wakeup
+  if they have been designated as wakeup sources using the standard
+  wakeup-source property, e.g.
```
Replaced with your suggestion

bjarki-trackunit commented on dts/bindings/power/atmel,sam-supc.yaml:
```diff
@@ -0,0 +1,47 @@
+
+    rtc: rtc@xxx {
+      ...
+      wakeup-source-id = <&supc SUPC_WAKEUP_SOURCE_RTT>;
```
Replaced with your suggestion

bjarki-trackunit commented:
@nandojve In progress :)

nandojve reviewed (approved):
LGTM!

gmarull reviewed (approved):
ok to poweroff changes
